### PR TITLE
Update environment-modules package

### LINKED ISSUE #10707: Update environment-modules to latest versions
As a user of Spack I want the `environment-modules` package to be updated so that I can install v4.X

### Rationale

Being able to install `environment-modules` v4.X is an essential part of integrating it into Spack!

### Additional information

Checking the documentation of [Environment Modules](https://modules.readthedocs.io/en/latest/) it seems that an initial list of modifications to the package should be:
- [ ] move the url to github
- [ ] add versions from 4.2.2 until 4.0.0
- [ ] check if new configure options (e.g. `--enable-auto-handling`, `--with-quarantine-vars`, `--with-pager`, `--with-pager-opts`, etc.) are to be used in the recipe

### PR DESCRIPTION
**Pull request by alalazo**
fixes #10707 

- [x] Changed url to point to github
- [x] Added versions up to 4.0.0 with sha256 checksums
- [x] Updated build recipe to account for new options and dependencies

### CODE REVIEW

**Comment by alalazo** [comment]
@xdelaruelle @RemiLacroix-IDRIS I started this draft PR to update `environemnt-modules`. and will ping you again when ready. In the meanwhile, if you have suggestions on how to build the package feel free to comment here.

File: var/spack/repos/builtin/packages/environment-modules/package.py [hunk 1]
```diff
@@ -3,18 +3,29 @@
...
 import os
 
 
 class EnvironmentModules(Package):
     """The Environment Modules package provides for the dynamic
     modification of a user's environment via modulefiles."""
 
-    homepage = "https://sourceforge.net/p/modules/wiki/Home/"
-    url = "http://prdownloads.sourceforge.net/modules/modules-3.2.10.tar.gz"
+    homepage = 'https://modules.readthedocs.io/en/latest/'
```
Comment (xdelaruelle) [inline var/spack/repos/builtin/packages/environment-modules/package.py]: I would suggest to use here a link to the website (`https://cea-hpc.github.io/modules/`) rather a link to the documentation portal.

File: var/spack/repos/builtin/packages/environment-modules/package.py [hunk 2]
```diff
@@ -3,64 +3,81 @@
...
+            # See: https://sourceforge.net/p/modules/bugs/62/
+            config_args.extend([
+                '--disable-debug',
+                'CPPFLAGS=-DUSE_INTERP_ERRORLINE'
+            ])
+
         configure(*config_args)
-        make()
+
+        targets = ['-C', 'doc', 'all'] if '@4.0.0:' else []
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/environment-modules/package.py]: @xdelaruelle Can you please double check the target I am building? Also, would it be fine with you if I add your name as a maintainer of the package :slightly_smiling_face:

**Comment by alalazo** [comment]
@xdelaruelle @RemiLacroix-IDRIS Do you think we'll need a variant to activate `--enable-auto-handling` from version `@4.2.2:` or should we just activate that by default if the option is present?

File: var/spack/repos/builtin/packages/environment-modules/package.py (hunk 2 above)
Comment (xdelaruelle) [reply var/spack/repos/builtin/packages/environment-modules/package.py]: A bare `make` call should be fine. Why do you add specific targets ?

Fine to get named as maintainer of the package :slightly_smiling_face:

Comment (alalazo) [reply var/spack/repos/builtin/packages/environment-modules/package.py]: > A bare make call should be fine. Why do you add specific targets ?

Following the advice in `README.md`. Do I need then the dependency on Sphinx?

Comment (xdelaruelle) [reply var/spack/repos/builtin/packages/environment-modules/package.py]: I see. README is unclear, I will clarify that. Installation instructions are available in https://github.com/cea-hpc/modules/blob/master/INSTALL.rst. The build is achieved with a basic `./configure; make; make install`.

I suggest to disable the build of the compatibility version (compilation of version 3.2 of Modules in addition of version 4) by adding the `--disable-compat-version` argument to the `./configure` script.

Requirements to build from a source tarball are:
* bash
* make
* sed

Then to run, Tcl >= 8.4 is needed.

Sphinx is not needed when building from a source tarball (only needed when building from the git repository)

**Comment by xdelaruelle** [comment]
@alalazo I would suggest to activate `--enable-auto-handling` by default if the option is present (starting 4.2.0)

**Comment by alalazo** [comment]
@xdelaruelle Let me know if you're fine with the package in the current state

File: var/spack/repos/builtin/packages/environment-modules/package.py [hunk 3]
```diff
@@ -3,64 +3,84 @@
...
         ]
 
         if '~X' in spec:
             config_args = ['--without-x'] + config_args
 
+        if '@4.2.0:' in self.spec:
+            config_args.extend([
+                '--enable-auto-handling'
+            ])
+
```
Comment (xdelaruelle) [inline var/spack/repos/builtin/packages/environment-modules/package.py]: Starting version 4.1.0, I would suggest to add the following `./configure` argument:

     --with-quarantine-vars=LD_LIBRARY_PATH

This setting enables [Modules quarantine mechanism](https://modules.readthedocs.io/en/stable/MIGRATING.html#quarantine-mechanism-to-protect-module-execution), which avoids being poisoned by an incompatible Tcl installation coming from one software bundle enabled in user environment.

**Comment by xdelaruelle** [comment]
@alalazo Apart from the quarantine mechanism point, things look good to me

**Comment by RemiLacroix-IDRIS** [comment]
> @xdelaruelle @RemiLacroix-IDRIS Do you think we'll need a variant to activate `--enable-auto-handling` from version `@4.2.2:` or should we just activate that by default if the option is present?

I wasn't aware of this option. I guess people who don't want to use this feature can always turn it off at runtime so I don't see anything wrong with enabling it by default.

File: var/spack/repos/builtin/packages/environment-modules/package.py (hunk 3 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/environment-modules/package.py]: Last request: can you double check the comment I wrote in 69d9dd8? Just to be sure my understanding is correct and that I won't put the next guy changing this package on a wrong track...

**Pull request merged**